src/ds/ref.h

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include <utility>
+
+template <typename T>
+class Ref final
```

Comment:
Why do we need a custom reference-counted container - can this be replaced by `std::shared_ptr`?

Reply:
My first attempt at this was with shared_ptr. The use of the atomics used in shared_ptr introduced a really high CPU overhead and make this not worthwhile

Reply:
Could `Ref` usage be replaced by a `unique_ptr` and have it passed between threads to avoid the copying? If this is shared between threads how do we guarantee that the internal state is not being modified by multiple threads?